### ISSUE TITLE
The object does not have an `id` if it is an instance of a Rails model in which the table is a Postgres view.

### ISSUE THREAD

**Issue by wrmk** [rails/rails#49950]
### Steps to reproduce
The object does not have an `id` if it is an instance of a Rails model in which the table is a Postgres view.

This worked in Rails v7.0.8 and fails in v7.1.1.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "~> 7.1.0"
  gem "pg", "~> 1.5.4"
end

require "active_record"
require "minitest/autorun"
require "logger"

ActiveRecord::Base.establish_connection(ENV.fetch("DATABASE_URL", "postgres:///"))
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Base.connection.exec_query(<<~SQL)
  create temporary table animals (
    id int primary key generated by default as identity,
    name text not null
  );
SQL

ActiveRecord::Base.connection.exec_query(<<~SQL)
  create temporary view dogs as SELECT animals.id, animals.name FROM animals;
SQL

class Dog < ActiveRecord::Base
  self.primary_key = :id
end

class BugTest < Minitest::Test
  def test_id_is_returned
    dog = Dog.create!(name: "Jack")
    assert(dog.id)
  end
end

```

### Expected behavior
`dog.id` should have the value from the database.

### Actual behavior
`dog.id` is `nil`

### System configuration
**Rails version**: 7.1.1

**Ruby version**: 3.2.2

**Comment by nvasilevski** [comment]
I was thinking about the issue and about the fix we are considering and wanted to dig a little bit further. It looks like at least in postgresql views won't have any indication of any auto-populated column, not just the primary key. For example if we tweak the schema a little bit:
```ruby
ActiveRecord::Base.connection.exec_query(<<~SQL)
  create temporary table animals (
    id int primary key generated by default as identity,
    name text not null,
    purchase_date DATE DEFAULT CURRENT_DATE
  );
SQL

ActiveRecord::Base.connection.exec_query(<<~SQL)
  create temporary view dogs as SELECT animals.id, animals.name, animals.purchase_date FROM animals;
SQL
```
The same assertion will fail for the `purchase_date`:
```ruby
    dog = Dog.create!(name: "Jack")
    assert(dog.purchase_date)
```

Basically because `Dog._returning_columns_for_insert` is an empty list.

So I think to simplify the fix we need to approach it in the following way:
Currently: We want to always append `primary_key` to `_returning_columns_for_insert` list
Possible new approach: Append `primary_key` to `_returning_columns_for_insert` only if the original `_returning_columns_for_insert` is empty, meaning that Rails wasn't able to find any value that needs to be auto-populated which means we should fallback to the previous behavior of returning PK.
I believe this should simplify the fix as currently it's a little tricky with composite primary key models and mysql not allowing to return anything other than a single value.


Ultimately I think we should try to make sure every autopopulated column from the view is returned but for now we can focus on PK to at least preserve the previous behavior.

@fatkodima let me know what you think and if you'd like to try this approach out. Also let me know if you prefer me doing it. I don't mind, just didn't want to take over this work since you've already come up with a possible fix. Thanks

**Comment by fatkodima** [comment]
@nvasilevski Yes, please try that approach, if you will. Do not currently have time to investigate it further.

**Comment by nvasilevski** [comment]
Should be fixed by https://github.com/rails/rails/pull/50783 and backported to `7-1-stable`
Please feel free to reopen if the fix doesn't work or it didn't cover your corner case. Thank you!

**Comment by wrmk** [comment]
The fix does not solve my problem and the test in the code example informs that id does not exist.
ruby 3.2.2
rails 7.1.0

**Comment by wrmk** [comment]
I can't reopen this issue, should I create a new one?

**Comment by nvasilevski** [comment]
Hey @wrmk, 

>rails 7.1.0

could you make sure you are testing against `main` or `7-1-stable`? The fix hasn't been released yet and will be available in Rails 7.1.4

I just ran the reproduction script against these versions:

```ruby
  gem "activerecord", github: "rails/rails", branch: "main"
```
and

```ruby
  gem "activerecord", github: "rails/rails", branch: "7-1-stable"
```

And it passes in both cases.

Feel free to open a new one as it will probably require details that are not present in the existing reproduction script to reveal a corner-case we might have missed. Thank you

**Comment by wrmk** [comment]
thanks, it solved my problem.
For some reason, I thought that fix would be applied for the 7.1 version on Rubygems
